### PR TITLE
Doxygen improvements to bilinearform

### PR DESCRIPTION
**Pull request by acfisher**
This is a small PR to add some improvements to the doxygen for bilinearform.hpp.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3972](https://github.com/mfem/mfem/pull/3972) | @acfisher | @tzanio | @tzanio + @termi-official + @v-dobrev + @mlstowell | 11/28/23 | 1/23/24 | 1/27/24 | |

<details>
<summary>PR Checklist</summary>
    
- [x] Code builds.
- [x] Code passes `make style`.
- [x] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [x] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [x] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [x] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [x] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [x] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [x] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [x] Update this checklist, if the new pull request affects it.
- [x] Run `make unittest` to make sure all unit tests pass.
- [x] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by acfisher** [comment]
@mlstowell as another person putting effort into the doxygen quality, do you want to buddy up and review each other's documentation work?  I plan to do a few of these little updates whenever I get a little time and they should be pretty straightforward to review.

File: fem/bilinearform.hpp [hunk 1]
```diff
@@ -324,14 +330,15 @@ class BilinearForm : public Matrix
...
    virtual MatrixInverse *Inverse() const;
 
-   /// Finalizes the matrix initialization.
+   /** @brief  Finalizes the matrix initialization if the ::AssemblyLevel is LEGACY.
+       THe matrix that gets finalized is different if you are using static condensation
+       or hybridization.*/
    virtual void Finalize(int skip_zeros = 1);
 
    /** @brief Returns a const reference to the sparse matrix:  \f$ M \f$
-
```
Comment (v-dobrev) [inline fem/bilinearform.hpp:334]: Empty lines like these have specific meaning -- they separate the brief section from the detailed section of the doxygen comments.

**Comment by termi-official** [comment]
xref https://github.com/mfem/mfem/pull/3688/ here, because I think it has it has gone under a bit

File: fem/bilinearform.hpp (hunk 1 above)
Comment (acfisher) [reply fem/bilinearform.hpp:334]: Ack, I'll put the spaces back in.  Thanks for catching that.

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Review by tzanio — APPROVED** [review]

File: fem/bilinearform.hpp [hunk 2]
```diff
@@ -324,14 +334,17 @@ class BilinearForm : public Matrix
    double InnerProduct(const Vector &x, const Vector &y) const
    { return mat->InnerProduct (x, y); }
 
-   /// Returns a pointer to (approximation) of the matrix inverse:  \f$ M^{-1} \f$
+   /// Returns a pointer to (approximation) of the matrix inverse:  \f$ M^{-1} \f$ (currently returns NULL)
```
[the remark this answers is not in the document]
Comment (acfisher) [reply fem/bilinearform.hpp:337]: Yup I took a look and it just returns a null

File: fem/bilinearform.hpp [hunk 3]
```diff
@@ -73,13 +73,16 @@ class BilinearForm : public Matrix
    /// FE space on which the form lives. Not owned.
    FiniteElementSpace *fes;
 
-   /// The assembly level of the form (full, partial, etc.)
+   /// The ::AssemblyLevel of the form (LEGACY, FULL, ELEMENT, PARTIAL)
```
[the remark this answers is not in the document]
Comment (acfisher) [reply fem/bilinearform.hpp:76]: No I don't know where to find it in the documntation for doxygen.  I found an example using the :: and I ran with it :).

**Comment by acfisher** [comment]
> I left some suggestions and comments since I am not sure about some points here.
> 
> I think the there is a problem on the docs for some functions, where internal details are documented rather than the actual functionality. But fixing this is rather future work than for this PR.

Thanks for reviewing this.  I liked all of your suggestions and pulled them in.  I'll merge master into it soon, and am hoping to get this pull request resolved soon.

File: fem/bilinearform.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (termi-official) [reply fem/bilinearform.hpp:337]: Sorry I am still confused. So this is just an unimplemented for now? If so, should we add this information to the doc string? Or is there some other reason which I am missing here?

**Comment by pazner** [comment]
Hi @termi-official, since #3688 is merged, could you take a quick look at this PR? Thanks!

File: fem/bilinearform.hpp [hunk 4]
```diff
@@ -121,18 +125,24 @@ class BilinearForm : public Matrix
...
 
    int precompute_sparsity;
-   // Allocate appropriate SparseMatrix and assign it to mat
+
+   /// Allocate appropriate SparseMatrix and assign it to #mat
    void AllocMat();
 
+   /** @brief For partially conforming trial and/or test FE spaces, complete the
+       assembly process by performing \f$ A := P^t A P\f$ where A is the internal
+       sparse matrix and P is the conforming prolongation matrice of the
```
Comment (termi-official) [inline fem/bilinearform.hpp:140]: I think A is here ambiguous for new users.

Comment (termi-official) [reply fem/bilinearform.hpp:140]:
```suggestion
       assembly process by performing \f$ A := P^t A P\f$ where \f$ A \f$ is the internal
       sparse matrix and \f$ P \f$ is the conforming prolongation matrix of the
```
plus some different symbol for a maybe?

File: fem/bilinearform.hpp [hunk 5]
```diff
@@ -324,14 +334,17 @@ class BilinearForm : public Matrix
    double InnerProduct(const Vector &x, const Vector &y) const
    { return mat->InnerProduct (x, y); }
 
-   /// Returns a pointer to (approximation) of the matrix inverse:  \f$ M^{-1} \f$
+   /// Returns a pointer to (approximation) of the matrix inverse:  \f$ M^{-1} \f$ (currently returns NULL)
    virtual MatrixInverse *Inverse() const;
 
-   /// Finalizes the matrix initialization.
+   /** @brief  Finalizes the matrix initialization if the ::AssemblyLevel is LEGACY.
```
Comment (termi-official) [inline fem/bilinearform.hpp:340]:
```suggestion
   /** @brief  Finalizes the matrix initialization if the ::AssemblyLevel is AssemblyLevel::LEGACY.
```

File: fem/bilinearform.hpp [hunk 6]
```diff
@@ -888,7 +920,7 @@ class MixedBilinearForm : public Matrix
    virtual const Operator *GetOutputRestriction() const
    { return test_fes->GetRestrictionMatrix(); }
 
-   /** For partially conforming trial and/or test FE spaces, complete the
+   /** @brief For partially conforming trial and/or test FE spaces, complete the
        assembly process by performing A := P2^t A P1 where A is the internal
```
Comment (termi-official) [inline fem/bilinearform.hpp:924]:
```suggestion
       assembly process by performing \f$ A := P2^t A P1 \f$ where A is the internal
```
I think the ambiguitiy of A might be confusing for new users here (as above)

**Review by termi-official — COMMENTED** [review]
I left some commens and together with the remainign comment above and this PR is good to go.

**Comment by pazner** [comment]
How about adding the following small documentation correction to this PR while we're at it:
```diff
diff --git a/fem/lininteg.hpp b/fem/lininteg.hpp
index 02fde00c9..c9b6e6dba 100644
--- a/fem/lininteg.hpp
+++ b/fem/lininteg.hpp
@@ -317,7 +317,7 @@ public:
 };
 
 /** Class for boundary integration of L(v) := (g, v), where
-    f=(f1,...,fn) and v=(v1,...,vn). */
+    g=(g1,...,gn) and v=(v1,...,vn). */
 class VectorBoundaryLFIntegrator : public LinearFormIntegrator
 {
 private:
```

File: fem/bilinearform.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (acfisher) [reply fem/bilinearform.hpp:337]: Yes it is not implemented at this time.  I'm not sure why it is there, but I wanted to let folks know that it doesn't do anything useful.

[the remark this answers is not in the document]
Comment (acfisher) [reply fem/bilinearform.hpp:337]: I changed it to mention that it is unimplemented.

File: fem/bilinearform.hpp [hunk 7]
```diff
@@ -121,18 +125,24 @@ class BilinearForm : public Matrix
...
    DiagonalPolicy diag_policy;
 
    int precompute_sparsity;
-   // Allocate appropriate SparseMatrix and assign it to mat
+
+   /// Allocate appropriate SparseMatrix and assign it to mat
    void AllocMat();
 
+   /** @brief For partially conforming trial and/or test FE spaces, complete the
+       assembly process by performing A := P^t A P where A is the internal
```
[the remark this answers is not in the document]
Comment (acfisher) [reply fem/bilinearform.hpp]: For these I removed the A from the left hand side which should make it more clear.

**Comment by acfisher** [comment]
> How about adding the following small documentation correction to this PR while we're at it:
> 
> ```diff
> diff --git a/fem/lininteg.hpp b/fem/lininteg.hpp
> index 02fde00c9..c9b6e6dba 100644
> --- a/fem/lininteg.hpp
> +++ b/fem/lininteg.hpp
> @@ -317,7 +317,7 @@ public:
>  };
>  
>  /** Class for boundary integration of L(v) := (g, v), where
> -    f=(f1,...,fn) and v=(v1,...,vn). */
> +    g=(g1,...,gn) and v=(v1,...,vn). */
>  class VectorBoundaryLFIntegrator : public LinearFormIntegrator
>  {
>  private:
> ```

Good catch I swept that change into this PR.

**Review by termi-official — APPROVED** [review]
LTGM

**Comment by acfisher** [comment]
> LGTM

Thanks for putting the time in to review and improve this PR.

**Comment by acfisher** [comment]
@tzanio I believe this is ready to go into next.

File: fem/bilinearform.hpp [hunk 8]
```diff
@@ -73,13 +73,16 @@ class BilinearForm : public Matrix
    /// FE space on which the form lives. Not owned.
    FiniteElementSpace *fes;
 
-   /// The assembly level of the form (full, partial, etc.)
+   /// The ::AssemblyLevel of the form (AssemblyLevel::LEGACY, AssemblyLevel::FULL, AssemblyLevel::ELEMENT, AssemblyLevel::PARTIAL)
```
Comment (tzanio) [inline fem/bilinearform.hpp:76]: Can we try to fit to 80 columns?

File: fem/bilinearform.hpp [hunk 9]
```diff
@@ -1009,18 +1051,22 @@ class MixedBilinearForm : public Matrix
...
    /// Read-only access to the associated trial FiniteElementSpace.
    const FiniteElementSpace *TrialFESpace() const { return trial_fes; }
 
    /// Return the test FE space associated with the BilinearForm.
    FiniteElementSpace *TestFESpace() { return test_fes; }
+
    /// Read-only access to the associated test FiniteElementSpace.
    const FiniteElementSpace *TestFESpace() const { return test_fes; }
 
+   /// Deletes internal matrices, bilinear integrators, and the BilinearFormExtension
```
Comment (tzanio) [inline fem/bilinearform.hpp:1069]: Can we try to fit to 80 columns?

**Comment by tzanio** [comment]
Thanks!

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Pull request merged**